doc/admin-guide/files/remap.config.en.rst

```diff
@@ -518,20 +514,95 @@ mapping rules. (It is activated before any mappings and is never
 deactivated.) The filter `local_only` will only be applied to the
 second mapping.
 
-Implict IPAllow filter
-======================
+Special Filter and ip_allow Named Filter
+----------------------------------------
+
+If :file:`ip_allow.yaml` has a "deny all" filter, it is treated as a special filter that is applied before remapping for
+optimizaion. To control this for specific remap rules, a named filter called `ip_allow` is pre-defined. This named filter is
+activated implicitly in default. To stop applying the special rule, disable the `ip_allow` filter as shown below.
+
+::
 
-To allow control of :ref:`IP Allow<ip-allow>` it is treated as an implicitly active and named
-filter. When this filter is active IP Allow checks are done before remap. To prevent this for
-specific remap rules, this filter, named "ip_allow", must be disabled. The common way of doing this
-would be ::
+   # ip_allow.yaml
+   ip_allow:
+      - apply: in
+        ip_addrs: 198.51.100.0/24
+        action: deny
+        method: ALL
 
+   # remap.config
    .deactivatefilter ip_allow
    map ...
    map ...
    .activatefilter ip_allow
 
-Note this entirely disables IP Allow checks for those remap rules.
+Note this entirely disables :file:`ip_allow.yaml` checks for those remap rules.
+
+Evaluation Order and Matching Policy
+------------------------------------
+
+ATS evaluates multiple ACL filters in the following order:
+
+1. Special "deny all" filter in :file:`ip_allow.yaml`
+2. In-line Filter in :file:`remap.config`
+3. Named Filter in :file:`remap.config`
+4. Filters in :file:`ip_allow.yaml`
+
+When an ACL filter is found, ATS stops processing ACL filters depending on the mathcing policy configured by
+:ts:cv:`proxy.config.url_remap.acl_matching_policy`.
+
+First Explicit Match Wins Policy
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This is the default matching policy. ATS stops processing ACL filters only if an explicit match of IP address and HTTP method is
```

Comment:
Sounds like we need to talk which config should be default. 

I'd prefer compatible with old ATS versions. New behavior as default seems too painful on upgrading to me.

Reply:
Adding the configuration shouldn't be too bad, right? There's already configuration change requirements involved with ATS 10. We changed the entire format of the file to YAML. :) I can't image setting this to the non-default value is unreasonable.

We need to rather decide which is the proper default for users and make that the default.

---

While writing this comment, the thought occurs to me that maybe we should take a slightly different approach to this. Maybe it makes sense for us to build on top of your configuration switch and make it work like this:

* **Legacy ACL matching.** This keeps ACL rule matching the way ATS 9 and prioror matches.
* **Enhanced ACL filters.** With this enabled, `@action=deny` and `@action=allow` function like ip_allow.yaml rules (i.e., the way my change makes them). But in addition to those actions, we can add `@action=add_deny` and `@action=add_allow`. These would allow users to add specific methods to deny or allow lists, like ATS 9 behaves currently behaves with `@action=deny` and `@action=allow`. This way enhanced offers a super set of functionality.

Going this route, we would make the "enhanced" feature the default because it is unambiguously superior. But since it is backwards incompatible, the legacy switch is added (non-default) to ease transition to ATS 10.

This will also make things easier for users to transition to the new behavior because remap.config would simply be changed with a few sed commands to transition `@action=deny` to `@action=add_deny` and `@action=allow` to `@action=add_allow`.

Reply:
I would suggest keeping the default behaviour the same as 9.x only because the human errors that could occur by changing the default are pretty high. Changing the default would make the assumption administrates will read the new docs thoroughly and update their configs to replicate the same behaviour. 

If we implicitly allow methods that are not specified, we are creating behaviour that is not visible to the administrators. They may not think like a software engineer and investigate in a way that any of us would which leaves big gaps of vulnerabilities that could open up. 

I think keeping the newer behaviour as an optional enhancement for the right people to try would make more sense. They would be updating their configs knowing the implications of the change rather than just trying to make it all run again.

Reply:
Thanks for the thoughts, @masaori335 and @jasmine-nahrain .

I would be content with:
1. Make it the default for ATS 10, but deprecate it.
2. Remove it for ATS 11

Does that sound fine?